File: Interview-17-12/Java/Solution.java
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

/**
 * https://leetcode.cn/problems/binode-lcci/
 */
public class Solution {

    TreeNode dummyNode = new TreeNode();
    TreeNode curr = dummyNode;
    
    public TreeNode convertBiNode(TreeNode root) {
        inorder(root);
        return dummyNode.right;
    }

    private void inorder(TreeNode root) {
        if (root == null) return;
        inorder(root.left);
        curr.right = root;
        curr = root;
        root.left = null;
        inorder(root.right);
    }

}

File: Interview-17-12/Python/Solution.py
# Definition for a binary tree node.
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def __init__(self):
        self.dummyNode = TreeNode()
        self.curr = self.dummyNode

    def convertBiNode(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        self.inorder(root)
        return self.dummyNode.right

    def inorder(self, root: TreeNode):
        if root is None:
            return
        self.inorder(root.left)

        self.curr.right = root
        self.curr = root
        root.left = None

        self.inorder(root.right)
